### PR TITLE
Additional Built In Rules for Azure API Management

### PR DESCRIPTION
**Pull request by harissohail55**
## Description

This change includes the addition of 7 new security rules for Azure API Management. The 7 rules are already available in Azure Policy, and this addition in the template-analyzer will ensure customers are secure when deploying via IaC as well.

**Existing Azure API Management rules (approved in a previous PR):**
TA-000029 - Azure API Management APIs should use HTTPS only

**New Azure API Management rules (in this PR):**
TA-000032: API Management calls to API backends should not bypass certificate thumbprint or name validation
TA-000033: API Management direct management endpoint should not be enabled
TA-000034: API Management minimum API version should be set to 2019-12-01 or higher
TA-000035: API Management Named Value secrets should be stored in Azure Key Vault
TA-000036: API Management services should use a virtual network
TA-000037: API Management subscriptions should not be scoped at the all API scope
TA-000038: API Management calls to API backends should be authenticated

---

This checklist is used to make sure that common guidelines for a pull request are followed.
- [X] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [X] The pull request does not introduce breaking changes.

### [General Guidelines](../CONTRIBUTING.md)
- [X] Title of the pull request is clear and informative.
- [X] Description of the pull request is clear and informative.
- [x] I have added myself to the 'assignees'.
- [X] I have added 'linked issues' if relevant.

### [Testing Guidelines](../CONTRIBUTING.md#tests)
- [X] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Comment by harissohail55** [comment]
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```

**Comment by harissohail55** [comment]
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```

@microsoft-github-policy-service agree company="Microsoft"

**Comment by harissohail55** [comment]
@microsoft-github-policy-service agree company="Microsoft"

File: docs/built-in-bpa-rules.md [hunk 1]
```diff
@@ -191,4 +191,46 @@ Azure supports two deployment models: Classic and Azure Resource Manager (ARM).
...
+### TA-000032: API Management calls to API backends should not bypass certificate thumbprint or name validation
+API Management calls to API backends should not bypass certificate thumbprint or name validation.
+
+**Recommendation**: To improve security, API Management should validate the backend server certificate for all API calls. Enable SSL certificate thumbprint and name validation by setting the *tls.validateCertificateName* and *tls.validateCertificateChain* properties to true
+#### Severity: 2
+
+### TA-000033: API Management direct management endpoint should not be enabled
+The direct management REST API in Azure API Management bypasses Azure Resource Manager role-based access control, authorization, and throttling mechanisms, thus increasing the vulnerability of your service.
+
+**Recommendation**: Set the *properties.enabled* property to false, to ensure direct access to the management REST API is disabled
```
Comment (anaismiller) [inline docs/built-in-bpa-rules.md:205]:
```suggestion
**Recommendation**: Set the *properties.enabled* property to `false` to ensure that direct access to the management REST API is disabled.
```

File: docs/built-in-bpa-rules.md [hunk 2]
```diff
@@ -191,4 +191,46 @@ Azure supports two deployment models: Classic and Azure Resource Manager (ARM).
...
+#### Severity: 3
+
+### TA-000034: API Management minimum API version should be set to 2019-12-01 or higher
+To prevent service secrets from being shared with read-only users, the minimum API version should be set to 2019-12-01 or higher.
+
+**Recommendation**: Set the *apiVersionContstraint.minAPIversion* property to 2019-12-01 or higher to prevent service secrets from being shared with read-only users.
+#### Severity: 2
+
+### TA-000035: API Management Named Value secrets should be stored in Azure Key Vault
```
Comment (anaismiller) [inline docs/built-in-bpa-rules.md:214]:
```suggestion
### TA-000035: API Management Named Values secrets should be stored in Azure Key Vault
```

File: docs/built-in-bpa-rules.md [hunk 3]
```diff
@@ -191,4 +191,46 @@ Azure supports two deployment models: Classic and Azure Resource Manager (ARM).
...
+Named values are a collection of name and value pairs in each API Management service. Secret values can be stored either as encrypted text in API Management (custom secrets) or by referencing secrets in Azure Key Vault. To improve security of API Management and secrets, reference secret named values from Azure Key Vault. Azure Key Vault supports granular access management and secret rotation policies.
```
Comment (anaismiller) [inline docs/built-in-bpa-rules.md:215]:
```suggestion
Named Values are a collection of name and value pairs in each API Management service. Secret values can be stored either as encrypted text in API Management (custom secrets) or by referencing secrets in Azure Key Vault. To improve security of API Management and secrets, reference secret Named Values from Azure Key Vault. Azure Key Vault supports granular access management and secret rotation policies.
```

File: docs/built-in-bpa-rules.md [hunk 4]
```diff
@@ -191,4 +191,46 @@ Azure supports two deployment models: Classic and Azure Resource Manager (ARM).
...
+
+**Recommendation**: To utilize secrets stored in Key Vault for Azure API Management, learn more [here](https://learn.microsoft.com/en-us/azure/api-management/api-management-howto-properties?tabs=azure-portal#key-vault-secrets).
```
Comment (anaismiller) [inline docs/built-in-bpa-rules.md:217]:
```suggestion
**Recommendation**: To utilize secrets stored in Key Vault for Azure API Management, learn more [here](https://learn.microsoft.com/azure/api-management/api-management-howto-properties?tabs=azure-portal#key-vault-secrets).
```

File: src/Analyzer.Core.BuiltInRuleTests/Tests/TA-000032/TA-000032.json [hunk 5]
```diff
@@ -0,0 +1,46 @@
+[
+  {
+    "Template": "APIMService.json",
+    "ReportedFailures": [
+      {
+        "LineNumber": 50,
+        "Description": "Validate Certificate Chain is False"
```
Comment (anaismiller) [inline src/Analyzer.Core.BuiltInRuleTests/Tests/TA-000032/TA-000032.json:7]: Is it usually `False` or `false` or does not matter?

File: docs/built-in-bpa-rules.md [hunk 6]
```diff
@@ -191,4 +191,46 @@ Azure supports two deployment models: Classic and Azure Resource Manager (ARM).
 Azure supports two deployment models: Classic and Azure Resource Manager (ARM). ARM provides several security enhancements to the Classic model. You can learn more about the differences [here](https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/deployment-models).
 
 **Recommendation**: Use ARM for your storage accounts to provide security enhancements such as: stronger access control (RBAC), better auditing, ARM-based deployment and governance, access to managed identities, access to key vault for secrets, Azure AD-based authentication and support for tags and resource groups for easier security management.
-#### Severity: 1
\ No newline at end of file
+#### Severity: 1
+
+### TA-000032: API Management calls to API backends should not bypass certificate thumbprint or name validation
+API Management calls to API backends should not bypass certificate thumbprint or name validation.
```
[the remark this answers is not in the document]
Comment (harissohail55) [reply docs/built-in-bpa-rules.md:199]: Hey Vera, yep, the Azure API Management team is in the process of updating the Azure Policy titles as well. This description is the updated text, that will soon reflect in the Azure Policy description as well (same case for other rules)

File: src/Analyzer.Core.BuiltInRuleTests/Tests/TA-000032/TA-000032.json (hunk 5 above)
Comment (harissohail55) [reply src/Analyzer.Core.BuiltInRuleTests/Tests/TA-000032/TA-000032.json]: Updated to false

File: docs/built-in-bpa-rules.md (hunk 4 above)
Comment (harissohail55) [reply docs/built-in-bpa-rules.md]: Thanks updated in both docs & src/Analyzer.Core/Rules/BuiltInRules.json

File: docs/built-in-bpa-rules.md (hunk 2 above)
Comment (harissohail55) [reply docs/built-in-bpa-rules.md]: Thanks, updated in both docs & src/Analyzer.Core/Rules/BuiltInRules.json

File: docs/built-in-bpa-rules.md (hunk 1 above)
Comment (harissohail55) [reply docs/built-in-bpa-rules.md]: Thanks, updated in both docs & src/Analyzer.Core/Rules/BuiltInRules.json

File: docs/built-in-bpa-rules.md [hunk 7]
```diff
@@ -191,4 +191,46 @@ Azure supports two deployment models: Classic and Azure Resource Manager (ARM).
...
+### TA-000037: API Management subscriptions should not be scoped at the all API scope
+API Management subscriptions should be scoped to a product or an individual API instead of all APIs, which could result in an excessive data exposure.
+
+**Recommendation**: To ensure the API Management service is not configured at the all API scope, please set the *properties.scope* property to a value other than */apis*. For example, */product/{productId}*, or */apis/{apiId}*.
+#### Severity: 2
+
+### TA-000038: API Management calls to API backends should be authenticated
+Calls from API Management to backends should use some form of authentication, whether via certificates or credentials. Does not apply to Service Fabric backends.
+
+**Recommendation**: To enable API backend authentication, configure authentication credentials in the query parameters, client certificate, or Authorization header. To learn more about configuring Authentication in Azure API Management, please visit [here](https://learn.microsoft.com/en-us/azure/api-management/authentication-authorization-overview).
```
[the remark this answers is not in the document]
Comment (harissohail55) [reply docs/built-in-bpa-rules.md]: Thanks, updated in both docs & src/Analyzer.Core/Rules/BuiltInRules.json

File: docs/built-in-bpa-rules.md [hunk 8]
```diff
@@ -191,4 +191,46 @@ Azure supports two deployment models: Classic and Azure Resource Manager (ARM).
...
+
+**Recommendation**: To utilize secrets stored in Key Vault for Azure API Management, learn more [here](https://learn.microsoft.com/en-us/azure/api-management/api-management-howto-properties?tabs=azure-portal#key-vault-secrets).
+#### Severity: 2
+
+### TA-000036: API Management services should use a virtual network
+Azure Virtual Network deployment provides enhanced security, isolation and allows you to place your API Management service in a non-internet routable network that you control access to. These networks can then be connected to your on-premises networks using various VPN technologies, which enables access to your backend services within the network and/or on-premises. The developer portal and API gateway, can be configured to be accessible either from the Internet or only within the virtual network.
+
+**Recommendation**: To learn more about Virtual Networks for Azure API Management, please visit [here](https://aka.ms/apim-vnet).To see an example on how to configure a VNet via Azure Policy, please visit [here](https://learn.microsoft.com/en-us/samples/azure/azure-quickstart-templates/api-management-create-with-internal-vnet-application-gateway/).
```
[the remark this answers is not in the document]
Comment (harissohail55) [reply docs/built-in-bpa-rules.md]: Thanks, updated in both docs & src/Analyzer.Core/Rules/BuiltInRules.json

File: src/Analyzer.Core/Rules/BuiltInRules.json [hunk 9]
```diff
@@ -872,5 +872,223 @@
...
+        }
+      ]
+    }
+  },
+  {
+    "id": "TA-000033",
+    "name": "APIM.DirectAPIManagementEndpointShouldNotBeEnabled",
+    "shortDescription": "API Management direct management endpoint should not be enabled",
+    "fullDescription": "The direct management REST API in Azure API Management bypasses Azure Resource Manager role-based access control, authorization, and throttling mechanisms, thus increasing the vulnerability of your service.",
+    "recommendation": "Set the properties.enabled field to `false` to ensure that direct access to the management REST API is disabled.",
```
Comment (VeraBE) [inline src/Analyzer.Core/Rules/BuiltInRules.json:903]:
```suggestion
    "recommendation": "Set the properties.enabled field to false to ensure that direct access to the management REST API is disabled.",
```

File: src/Analyzer.Core/Rules/BuiltInRules.json [hunk 10]
```diff
@@ -872,5 +872,223 @@
       "path": "type",
       "notEquals": "Microsoft.ClassicStorage/storageAccounts"
     }
+  },
+  {
+    "id": "TA-000032",
+    "name": "APIM.BypassCertificateThumbprintOrNameValidation",
+    "shortDescription": "API Management calls to API backends should not bypass certificate thumbprint or name validation",
+    "fullDescription": "API Management calls to API backends should not bypass certificate thumbprint or name validation",
+    "recommendation": "To improve security, API Management should validate the backend server certificate for all API calls. Enable SSL certificate thumbprint and name validation by setting the tls.validateCertificateName and tls.validateCertificateChain properties to `true`.",
```
Comment (VeraBE) [inline src/Analyzer.Core/Rules/BuiltInRules.json:881]:
```suggestion
    "recommendation": "To improve security, API Management should validate the backend server certificate for all API calls. Enable SSL certificate thumbprint and name validation by setting the tls.validateCertificateName and tls.validateCertificateChain properties to true.",
```

File: src/Analyzer.Core/Rules/BuiltInRules.json (hunk 9 above)
Comment (harissohail55) [reply src/Analyzer.Core/Rules/BuiltInRules.json:903]: Hey Vera, in the previous PR comments, I was asked to put `` for property fields. I had updated based on the PR comments. Here's an example comment: https://github.com/Azure/template-analyzer/pull/316#discussion_r1093823805

File: src/Analyzer.Core/Rules/BuiltInRules.json (hunk 10 above)
Comment (harissohail55) [reply src/Analyzer.Core/Rules/BuiltInRules.json:881]: Hey Vera, same here. Should I use `` when referencing the properties?: https://github.com/Azure/template-analyzer/pull/316#discussion_r1093823805

File: src/Analyzer.Core/Rules/BuiltInRules.json [hunk 11]
```diff
@@ -872,5 +872,223 @@
...
+              "path": "properties.credentials.authorization.scheme",
+              "exists": true
+            },
+            {
+              "path": "properties.credentials.authorization.parameter",
+              "exists": true
+            }
+          ]
+        }
+
```
Comment (VeraBE) [inline src/Analyzer.Core/Rules/BuiltInRules.json:1090]:
```suggestion
```

**Review by VeraBE — APPROVED** [review]

File: docs/built-in-rules.md [hunk 12]
```diff
@@ -193,4 +193,46 @@ Azure supports two deployment models: Classic and Azure Resource Manager (ARM).
...
+#### Severity: 1
+
+### TA-000032: API Management calls to API backends should not bypass certificate thumbprint or name validation
+API Management calls to API backends should not bypass certificate thumbprint or name validation.
+
+**Recommendation**: To improve security, API Management should validate the backend server certificate for all API calls. Enable SSL certificate thumbprint and name validation by setting the *tls.validateCertificateName* and *tls.validateCertificateChain* properties to `true`.
+#### Severity: 2
+
+### TA-000033: API Management direct management endpoint should not be enabled
+The direct management REST API in Azure API Management bypasses Azure Resource Manager role-based access control, authorization, and throttling mechanisms, thus increasing the vulnerability of your service.
```
Comment (anaismiller) [inline docs/built-in-rules.md:205]:
```suggestion
The direct management REST API in Azure API Management bypasses Azure Resource Manager (ARM) Role-Based Access Control (RBAC), authorization, and throttling mechanisms, thus increasing the vulnerability of your service.
```

Comment (anaismiller) [reply docs/built-in-rules.md:205]: (trying to make it look more uniform to other rules in this docu

File: src/Analyzer.Core/Rules/BuiltInRules.json [hunk 13]
```diff
@@ -872,5 +872,222 @@
...
+          "equals": true
+        }
+      ]
+    }
+  },
+  {
+    "id": "TA-000033",
+    "name": "APIM.DirectAPIManagementEndpointShouldNotBeEnabled",
+    "shortDescription": "API Management direct management endpoint should not be enabled",
+    "fullDescription": "The direct management REST API in Azure API Management bypasses Azure Resource Manager role-based access control, authorization, and throttling mechanisms, thus increasing the vulnerability of your service.",
```
Comment (anaismiller) [inline src/Analyzer.Core/Rules/BuiltInRules.json:902]:
```suggestion
    "fullDescription": "The direct management REST API in Azure API Management bypasses Azure Resource Manager (ARM) Role-Based Access Control (RBAC), authorization, and throttling mechanisms, thus increasing the vulnerability of your service.",
```

File: docs/built-in-rules.md [hunk 14]
```diff
@@ -193,4 +193,46 @@ Azure supports two deployment models: Classic and Azure Resource Manager (ARM).
...
+#### Severity: 2
+
+### TA-000037: API Management subscriptions should not be scoped at the all API scope
+API Management subscriptions should be scoped to a product or an individual API instead of all APIs, which could result in an excessive data exposure.
+
+**Recommendation**: To ensure the API Management service is not configured at the all API scope, please set the *properties.scope* property to a value other than */apis*. For example, */product/{productId}*, or */apis/{apiId}*.
+#### Severity: 2
+
+### TA-000038: API Management calls to API backends should be authenticated
+Calls from API Management to backends should use some form of authentication, whether via certificates or credentials. Does not apply to Service Fabric backends.
```
Comment (anaismiller) [inline docs/built-in-rules.md:235]: Do you have a test case for `Does not apply to Service Fabric backends.`? If not, can you add one please?

Comment (harissohail55) [reply docs/built-in-rules.md:235]: Thanks Anais! I've verified with the APIM team on this, I've added a new check in the BuiltInRules for this case, and a new test case. I will also work with the APIM team to update the runtime policy to include this check as well.

**Review by anaismiller — APPROVED** [review]

**Pull request merged**